### DISCUSSION TITLE
variation yaw

### DISCUSSION THREAD

**Discussion by RealMadridFK** [rodralez/NaveGo#72] [General]
Hello, I have not seen it in the code. How the change in direction of movement is taken into account. Based on gyroscope readings only? And if I move forward, stop and ride backwards? Nav.yaw does not take into account such a situation?

**Comment by wangfeng35** [comment]
The accelerometer and gyroscope readings will detect the movement. As for your question, the estimated velocity and the position will determine your direction, even though the yaw did not change.

**Reply by RealMadridFK** [reply]
I just need to know the ground speed and its direction. I calculate the path speed by the sqrt(Vn^2 + Ve^2), therefore it will always be positive regardless of whether I am going forward or backward.

**Reply by wangfeng35** [reply]
The yaw angle is between -pi and pi, therefore the angle of going forward and backward is not the same. By comparing the yaw with the path direction, you will get the correct running direction.

**Reply by RealMadridFK** [reply]
![yaw](https://user-images.githubusercontent.com/90841994/139044155-066f9df2-8956-4c2d-8865-aff6eecf8e40.PNG)
![track](https://user-images.githubusercontent.com/90841994/139044158-44464b8e-db8d-423c-b661-fa33f658780c.PNG)
Track plot for GNSS and Navego. And a graph of the yaw change. The movement took place: back. Stop. Forward. Stop. Back.

**Reply by RealMadridFK** [reply]
![image](https://user-images.githubusercontent.com/90841994/139044826-bf1628c4-d26b-43ed-b455-cde570f04e4a.png)
For the algorithm to work adequately, I had to manually set this parameter like this. Tell me what it influences. If you leave it the way Allan thought, the trajectory becomes completely different.

**Reply by rodralez** [reply]
The ab_psd vector (1x3 accrs dynamic biases PSD) is part of the matrix Q. You are adjusting the level of the process noise of your system model, which sets the level of confidence that you have in you system model.

**Comment by rodralez** [comment]
If a vehicle moves forward, stops and goes backwards, the yaw angle has to change, definitely. The problem with the yaw angle is when the vehicle tracks a perfect straight line since the yaw angle starts being unobservable.

**Reply by RealMadridFK** [reply]
I mean, it would be nice to add some kind of accounting for whether we are moving forward or backward. That is, the yaw should not only depend on the readings of the gyroscopes, but somehow be supported by GNSS.

**Reply by rodralez** [reply]
Well, I was thinking... if the vehicle starts going backward without rotating 180 degrees, yes, the yaw angle isn't going to change its value. On the other hand, at least North velocity or East velocity has to change.

**Comment by RealMadridFK** [comment]
I don't understand how yet. But for example we are moving 40 degrees in the yaw angle. Then our speed changed signs. And we began to move along the opposite trajectory, without turning the vehicle, the yaw angle should change by 180 degrees.

**Reply by marcoamm** [reply]
I don't think the yaw angle has to change in this case. The definition of the yaw angle is given by the angle between the navigation frame and the body frame. If the vehicle is moving backward, the body frame hasn't rotated as it is fixed to the body. This should be accounted for only by the negative velocity in the body frame. If the geometry is properly modeled, the negative velocity will project the vehicle's position in the right direction.

**Reply by wangfeng35** [reply]
What's the velocity look like? The yaw should change with the running direction, but it's unreliable when the speed is low.

**Reply by wangfeng35** [reply]
By the way, the GNSS yaw angle is output in GPRMC sentence.
